### PR TITLE
Add a mirror to registry for testing images from branch 

### PR DESCRIPTION
**Pull request by JanZerebecki**
## Why is this PR needed?

Add --registry-mirror for testrunner

Fixes https://github.com/SUSE/avant-garde/issues/1601

## What does this PR do?

testrunner: Add --registry-mirror
    
to add a mirror to a registy that takes priority and can be used to
branched images. Creates a configuration that skuba then deploys to
/etc/containers/registries.conf .

--

Also deploy /etc/containers/registries.conf

Needed for https://github.com/SUSE/avant-garde/issues/1601 to be able to
remap registries to test branched images.

--

testrunner: Add config files to gather_logs

## Info for QA

I tested the following:
* an invalid registry location `--registry-replace https://example.com https://test.example.com` makes crio fail to start up and report this in its log
* an [unused registry location ](https://ci.suse.de/blue/organizations/jenkins/caasp-jobs%2Fpr%2Ftest/detail/PR-1127/26/pipeline/)`--registry-mirror example.com/test test.example.com/test` still works
* [adding to caspv5 a name that will fail](https://ci.suse.de/blue/organizations/jenkins/caasp-jobs%2Fpr%2Ftest/detail/PR-1127/27/pipeline) `--registry-mirror registry.suse.de/devel/caasp/5/containers/cr/containers/caasp/v5 does-not-exist.suse.de/caasp/v5` still works
* [adding to caspv5 a registry that has one image](https://ci.suse.de/blue/organizations/jenkins/caasp-jobs%2Fpr%2Ftest/detail/PR-1127/29/pipeline/) `--registry-mirror registry.suse.de/devel/caasp/5/containers/cr/containers/caasp/v5 registry.suse.de/home/jzerebecki/branches/devel/caasp/5/containers/cr/containers/caasp/v5` still works TODO the CI logs do not show which image was used, so if the right images was picked was not verified

### Status **AFTER** applying the patch

Test by using the features: --registry-replace for testrunner or addons/containers/registries.conf for skuba.

## Docs

https://confluence.suse.com/display/CaaSP/Jenkins+CI#JenkinsCI-Testbranchedimagesorpackagesthatgetincludedinimages

This PR adds support for a addons/containers directory. The existing skuba addons directory is currently not document: https://github.com/SUSE/doc-caasp/blob/master/adoc/deployment-bootstrap.adoc https://documentation.suse.com/suse-caasp/4.2/single-html/caasp-deployment/#bootstrap

The documentation about /etc/containers/registries.conf will still work as is, but it could mention that it is now possible to deploy it via skuba: https://github.com/SUSE/doc-caasp/blob/master/adoc/admin-crio-registries.adoc https://documentation.suse.com/suse-caasp/4.2/single-html/caasp-admin/#_configuring_container_registries_for_cri_o

If docs need to be updated, please add a link to a PR to https://github.com/SUSE/doc-caasp.
At the time of creating the issue, this PR can be work in progress (set its title to [WIP]),
but the documentation needs to be finalized before the PR can be merged. 

# Merge restrictions

(Please do not edit this)

We are in *v4-maintenance phase*, so we will restrict what can be merged to prevent unexpected surprises:

    What can be merged (merge criteria):
        2 approvals:
            1 developer: code is fine
            1 QA: QA is fine
        there is a PR for updating documentation (or a statement that this is not needed)

### CODE REVIEW

**Comment by JanZerebecki** [comment]
ready for review

File: internal/pkg/skuba/deployments/ssh/cri.go [hunk 1]
```diff
@@ -42,7 +43,7 @@ func criConfigure(t *Target, data interface{}) error {
 		if err != nil {
 			// If the deferred function has any return values, they are discarded when the function completes
 			// https://golang.org/ref/spec#Defer_statements
-			fmt.Println("Could not delete the crio.conf.d config path")
+			fmt.Println("Could not delete the path /tmp/crio.conf.d")
```
Comment (pablochacin) [inline internal/pkg/skuba/deployments/ssh/cri.go:46]: This is unrelated to the PR, right?

File: ci/infra/testrunner/testrunner.py [hunk 2]
```diff
@@ -171,6 +173,9 @@ def main():
                                help="Use cloud provider integration")
     bootstrap_args.add_argument("-t", "--timeout", type=int, default=300,
                                 help="timeout for waiting a node to become ready (seconds)")
+    bootstrap_args.add_argument("-m", "--registry-mirror", metavar='L',
+                                help="Add to the registry at the first L a mirror at the second L",
+                                nargs=2, action='append')
```
Comment (pablochacin) [inline ci/infra/testrunner/testrunner.py:178]: Maybe an example of how to use the parameter. I understand it is `--registry-mirror name http://url/to/registry`

Comment (JanZerebecki) [reply ci/infra/testrunner/testrunner.py:179]: done

**Comment by pablochacin** [comment]
LGTM. Let's wait for the feedback from the other squads.

File: ci/infra/testrunner/README.md [hunk 3]
```diff
@@ -463,7 +463,10 @@ optional arguments:
   -c, --cloud-provider  Use cloud provider integration
   -t TIMEOUT, --timeout TIMEOUT
                         timeout for waiting a node to become ready (seconds)
-
+  -m L L, --registry-mirror L L
+                        Add to the registry at the first L a mirror at the
+                        second L. Example:
```
Comment (rhafer) [inline ci/infra/testrunner/README.md:468]: Maybe it's just me, but this reads a bit weird. From this help string I have a hard time to understand what the option is really doing.

File: ci/infra/testrunner/skuba/skuba.py [hunk 4]
```diff
@@ -89,6 +94,26 @@ def node_bootstrap(self, cloud_provider=None, timeout=None):
 
         self.checker.check_node("master", 0, stage="joined", timeout=timeout)
 
+    def _setup_container_registries(self, registry_mirror):
+        mirrors = {}
+        for l in registry_mirror:
+            if l[0] not in mirrors:
+                mirrors[l[0]] = []
+            mirrors[l[0]].append(l[1])
+
```
Comment (rhafer) [inline ci/infra/testrunner/skuba/skuba.py:103]: So this implies that I can add multiple mirrors for a single registry, right? I am not sure how useful this is, but should it at least be documented in the help?

File: ci/infra/testrunner/README.md (hunk 3 above)
Comment (pablochacin) [reply ci/infra/testrunner/README.md:468]: Agree it should maybe read something like `-m prefix mirror`

File: ci/infra/testrunner/skuba/skuba.py (hunk 4 above)
Comment (pablochacin) [reply ci/infra/testrunner/skuba/skuba.py:103]: My understanding is that is possible in the cri-o configuration, but it is not actually implemented in this PR (maybe we could implement in a future PR allowing multple occurrences of the `--registry-mirror` parameter.

Comment (JanZerebecki) [reply ci/infra/testrunner/skuba/skuba.py:103]: AFAIK it will work in this PR. But I haven't tested it.

File: ci/infra/testrunner/README.md (hunk 3 above)
Comment (JanZerebecki) [reply ci/infra/testrunner/README.md:468]: How do I do that with the option library we use here? AFAIK it only supports one argument placeholder even if there are two.

Comment (pablochacin) [reply ci/infra/testrunner/README.md:468]: https://docs.python.org/3/library/argparse.html#metavar
```
parser = argparse.ArgumentParser(prog='PROG')
parser.add_argument('-x', nargs=2)
parser.add_argument('--foo', nargs=2, metavar=('bar', 'baz'))
parser.print_help()
usage: PROG [-h] [-x X X] [--foo bar baz]

optional arguments:
 -h, --help     show this help message and exit
 -x X X
 --foo bar baz
```

**Comment by dirkmueller** [comment]
@JanZerebecki could you please rebase the conflict?

File: ci/infra/testrunner/skuba/skuba.py (hunk 4 above)
Comment (JanZerebecki) [reply ci/infra/testrunner/skuba/skuba.py:98]: Done.

File: ci/infra/testrunner/README.md (hunk 3 above)
Comment (JanZerebecki) [reply ci/infra/testrunner/README.md]: Done. Is it now easy enough to understand?

Comment (rhafer) [reply ci/infra/testrunner/README.md]: Yes, thanks!

File: ci/infra/testrunner/README.md [hunk 5]
```diff
@@ -463,7 +463,13 @@ optional arguments:
   -c, --cloud-provider  Use cloud provider integration
   -t TIMEOUT, --timeout TIMEOUT
                         timeout for waiting a node to become ready (seconds)
-
+  -m R M, --registry-mirror R M
+                        Add to the registry R a mirror M. If an image is
+			available at the mirror it will be preferred, otherwise
```
Comment (rhafer) [inline ci/infra/testrunner/README.md:468]: indentation issue
